lib/util/glob-util.js

```diff
@@ -160,10 +154,25 @@ function listFilesToProcess(globPatterns, options) {
         const file = path.resolve(cwd, pattern);
 
         if (shell.test("-f", file)) {
-            addFile(fs.realpathSync(file), !shell.test("-d", file));
+            const ignoredPaths = new IgnoredPaths(options);
+
+            addFile(fs.realpathSync(file), !shell.test("-d", file), ignoredPaths);
         } else {
+
+            // regex to find .hidden or /.hidden patterns, but not ./relative or ../relative
+            const globIncludesDotfiles = /(?:(?:^\.)|(?:[\/\\]\.))[^\/\\\.].*/.test(pattern);
+
+            options.dotfiles = options.dotfiles || globIncludesDotfiles;
```

Comment:
As far as I can tell, it's internal only.  It's been there since we switched over to `node-glob`, in fact.  (https://github.com/eslint/eslint/blob/master/lib/ignored-paths.js#L113)